File: workers/common/worker_common/geometry/units.py
"""
DXF drawing-unit normalization ($INSUNITS).

The canonical internal unit is ALWAYS millimeters. Files authored in other
units are scaled at import time — decompose first (block INSERTs resolved
into flat primitives), then a uniform scale — so every downstream constant
(tolerances, channel widths, engine config) keeps its mm semantics.

$MEASUREMENT only toggles dimension-text units in CAD — ignored here.

Lot 2b (`docs/PLAN-IMPORT-2026-09-09.md` §9.2) : la table couvre les codes
**0 à 20** (sept codes seulement avant : un dessin déclaré en kilomètres ou
en yards était lu ×1, en silence, donc au millimètre), le code écrit en
**flottant** est accepté (notre propre exporteur DXF écrivait « 4.0 »), et
plus aucune unité n'est supposée sans **constat** (`insunits_detail`).

Les facteurs sont EXACTS par définition (le pouce VAUT 25,4 mm), pas ceux de
`ezdxf.units.METER_FACTOR` qui sont arrondis (1000/39.37007874 =
25,400000000101603). Le miroir Rust (`nest-import::units`) porte les mêmes
valeurs et les mêmes noms — c'est la parité des deux importeurs qui compte.
"""

from worker_common.logger import setup_logger

logger = setup_logger("units")

# $INSUNITS code -> multiplication factor to millimeters (exact).
INSUNITS_TO_MM = {
    1: 25.4,                        # inches
    2: 304.8,                       # feet
    3: 1609344.0,                   # miles
    4: 1.0,                         # millimeters
    5: 10.0,                        # centimeters
    6: 1000.0,                      # meters
    7: 1.0e6,                       # kilometers
    8: 2.54e-5,                     # microinches
    9: 0.0254,                      # mils (thousandths of an inch)
    10: 914.4,                      # yards
    11: 1.0e-7,                     # angstroms
    12: 1.0e-6,                     # nanometers
    13: 1.0e-3,                     # microns
    14: 100.0,                      # decimeters
    15: 10000.0,                    # decameters
    16: 100000.0,                   # hectometers
    17: 1.0e12,                     # gigameters
    18: 1.495978707e14,             # astronomical units (SI exact)
    19: 9.4607304725808e18,         # light years (SI exact)
    20: 3.0856775814913673e19,      # parsecs (SI exact)
}

# Identifiant canonique de l'unité — MÊMES chaînes que nest-import::units
# (le verrou de parité compare `unitDetected` entre les deux importeurs).
INSUNITS_NAME = {
    0: "unitless",
    1: "inch",
    2: "foot",
    3: "mile",
    4: "mm",
    5: "cm",
    6: "m",
    7: "km",
    8: "microinch",
    9: "mil",
    10: "yard",
    11: "angstrom",
    12: "nanometer",
    13: "micron",
    14: "decimeter",
    15: "decameter",
    16: "hectometer",
    17: "gigameter",
    18: "au",
    19: "lightyear",
    20: "parsec",
}

INSUNITS_MM = 4
MEASUREMENT_METRIC = 1

MM_PER_INCH = 25.4

# Au-delà de ce facteur (10 m par unité), l'unité déclarée ne peut pas être
# celle d'une pièce de tôlerie : la conversion reste EXACTE (jamais un repli
# silencieux sur le millimètre), mais elle est annoncée.
IMPLAUSIBLE_FACTOR_MM = 1.0e4


def output_scale_and_headers(output_unit: str):
    """
    (scale factor FROM mm, $INSUNITS, $MEASUREMENT) for result DXF export.

    Geometry is computed in canonical mm and converted only at this export
    boundary, full precision — never round geometry. Inches get
    $MEASUREMENT=0 (imperial) so both headers agree with the numbers.
    Unknown units fall back to mm (safe default for legacy jobs).
    """
    if output_unit == "inch":
        return 1.0 / MM_PER_INCH, 1, 0
    return 1.0, INSUNITS_MM, MEASUREMENT_METRIC


def unit_name(code: int) -> str:
    """Identifiant canonique de l'unité d'un code $INSUNITS."""
    return INSUNITS_NAME.get(code, "unknown")


def insunits_code(doc) -> int:
    """Declared $INSUNITS of a DXF document (0 = unitless / missing).

    TOLÉRANT AU FLOTTANT (lot 2b) : le code 70 est un entier par la spec,
    mais des producteurs l'écrivent « 4.0 » — dont notre propre exporteur DXF
    jusqu'au lot 2b. `int("4.0")` lève, et le repli valait « sans unité ».
    """
    try:
        raw = doc.header.get("$INSUNITS", 0)
    except Exception:
        return 0
    if raw is None or raw == "":
        return 0
    try:
        return int(raw)
    except (TypeError, ValueError):
        pass
    try:
        return int(round(float(str(raw).strip())))
    except (TypeError, ValueError):
        return 0


def insunits_detail(doc):
    """(code, nom canonique, facteur vers mm, constats) — lot 2b.

    Les textes des constats sont IDENTIQUES à ceux du miroir Rust
    (`nest-import::dxf::canonical`) : les deux importeurs disent la même
    chose du même fichier. La mise en forme utilisateur (FR/EN, niveau
    d'alerte) est le lot 2c.
    """
    code = insunits_code(doc)
    warnings = []
    if code in (0, INSUNITS_MM):
        factor = 1.0
        if code == 0:
            warnings.append("$INSUNITS missing or 0 — assuming millimeters")
        return code, unit_name(code), factor, warnings
    factor = INSUNITS_TO_MM.get(code)
    if factor is None:
        warnings.append(f"unknown $INSUNITS={code} — assuming millimeters")
        return code, unit_name(code), 1.0, warnings
    if factor >= IMPLAUSIBLE_FACTOR_MM:
        warnings.append(
            f"$INSUNITS={code} ({unit_name(code)}) — geometry scaled "
            f"x{factor:g} to mm"
        )
    return code, unit_name(code), factor, warnings


def insunits_to_mm(doc) -> float:
    """
    Multiplication factor converting the document's drawing units to mm.
    Unitless (0) or unknown codes fall back to 1.0 — i.e. treated as mm,
    which is the historical behaviour for every file uploaded so far.
    """
    code, name, factor, warnings = insunits_detail(doc)
    for message in warnings:
        logger.warning(message)
    if factor != 1.0:
        logger.info(
            f"DXF declared $INSUNITS={code} ({name}) -> scaling geometry "
            f"x{factor:g} to mm."
        )
    return factor

File: workers/geometry/crates/nest-import/src/units.rs
//! $INSUNITS normalization — canonical mm (worker_common/geometry/units.py twin).
//!
//! Lot 2b (`docs/PLAN-IMPORT-2026-09-09.md` §9.2) : table COMPLÈTE des codes
//! 0 à 20. Avant ce lot, sept codes seulement étaient connus — un dessin
//! déclaré en kilomètres (code 7) ou en yards (10) était lu **×1**, en
//! silence, donc à l'échelle du millimètre.
//!
//! Les facteurs sont **exacts par définition** (le pouce VAUT 25,4 mm), pas
//! ceux de `ezdxf.units.METER_FACTOR` qui sont arrondis : `1000/39.37007874`
//! rend 25,400000000101603 au lieu de 25,4. Le miroir Python porte les mêmes
//! valeurs — c'est la parité qui compte, pas la source.

/// $INSUNITS code → multiplication factor to millimeters (exact).
/// `None` = code inconnu (au-delà de 20, ou négatif) et 0 = sans unité :
/// les deux cas sont traités par `factor_to_mm`.
pub fn insunits_factor(code: i32) -> Option<f64> {
    match code {
        1 => Some(25.4),                        // inches
        2 => Some(304.8),                       // feet
        3 => Some(1_609_344.0),                 // miles
        4 => Some(1.0),                         // millimeters
        5 => Some(10.0),                        // centimeters
        6 => Some(1000.0),                      // meters
        7 => Some(1.0e6),                       // kilometers
        8 => Some(2.54e-5),                     // microinches
        9 => Some(0.0254),                      // mils
        10 => Some(914.4),                      // yards
        11 => Some(1.0e-7),                     // angstroms
        12 => Some(1.0e-6),                     // nanometers
        13 => Some(1.0e-3),                     // microns
        14 => Some(100.0),                      // decimeters
        15 => Some(10_000.0),                   // decameters
        16 => Some(100_000.0),                  // hectometers
        17 => Some(1.0e12),                     // gigameters
        18 => Some(1.495_978_707e14),           // astronomical units (SI exact)
        19 => Some(9.460_730_472_580_8e18),     // light years (SI exact)
        20 => Some(3.085_677_581_491_367_3e19), // parsecs (SI exact)
        _ => None,
    }
}

/// Identifiant canonique de l'unité — MÊME chaîne des deux côtés (le verrou
/// de parité compare `unitDetected` entre les deux importeurs).
pub fn unit_name(code: i32) -> &'static str {
    match code {
        0 => "unitless",
        1 => "inch",
        2 => "foot",
        3 => "mile",
        4 => "mm",
        5 => "cm",
        6 => "m",
        7 => "km",
        8 => "microinch",
        9 => "mil",
        10 => "yard",
        11 => "angstrom",
        12 => "nanometer",
        13 => "micron",
        14 => "decimeter",
        15 => "decameter",
        16 => "hectometer",
        17 => "gigameter",
        18 => "au",
        19 => "lightyear",
        20 => "parsec",
        _ => "unknown",
    }
}

/// Au-delà de ce facteur (10 m par unité), l'unité déclarée ne peut pas être
/// celle d'une pièce de tôlerie : la conversion reste EXACTE (jamais un repli
/// silencieux sur le millimètre), mais elle est annoncée — sinon l'utilisateur
/// reçoit une pièce de trois kilomètres sans savoir pourquoi.
pub const IMPLAUSIBLE_FACTOR_MM: f64 = 1.0e4;

/// Unitless (0) / mm (4) → 1.0 sans avertissement. Code connu → son facteur
/// exact. Code inconnu → 1.0 + drapeau d'avertissement (comportement
/// historique : mm supposés).
pub fn factor_to_mm(code: i32) -> (f64, bool) {
    if code == 0 || code == 4 {
        return (1.0, false);
    }
    match insunits_factor(code) {
        Some(f) => (f, false),
        None => (1.0, true), // unknown → warn
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn known_factors_are_exact() {
        // Jumeau de worker_common/geometry/units.py — valeurs figées.
        assert_eq!(insunits_factor(1), Some(25.4)); // inches
        assert_eq!(insunits_factor(2), Some(304.8)); // feet
        assert_eq!(insunits_factor(4), Some(1.0)); // mm
        assert_eq!(insunits_factor(5), Some(10.0)); // cm
        assert_eq!(insunits_factor(6), Some(1000.0)); // meters
        assert_eq!(insunits_factor(8), Some(2.54e-5)); // microinches
        assert_eq!(insunits_factor(9), Some(0.0254)); // mils
    }

    /// Lot 2b : les codes qui étaient lus ×1 en silence.
    #[test]
    fn the_whole_table_zero_to_twenty_is_covered() {
        for code in 1..=20 {
            assert!(
                insunits_factor(code).is_some(),
                "code {code} sans facteur — c'est un ×1 silencieux"
            );
            assert_ne!(unit_name(code), "unknown", "code {code} sans nom");
        }
        // Les facteurs exacts que le silence coûtait le plus cher :
        assert_eq!(insunits_factor(3), Some(1_609_344.0)); // mile
        assert_eq!(insunits_factor(7), Some(1.0e6)); // km
        assert_eq!(insunits_factor(10), Some(914.4)); // yard
        assert_eq!(insunits_factor(13), Some(1.0e-3)); // micron
        assert_eq!(insunits_factor(14), Some(100.0)); // dm
        // Ces facteurs sont EXACTS, pas ceux de ezdxf (arrondis) :
        assert_ne!(insunits_factor(1), Some(1000.0 / 39.37007874));
    }

    #[test]
    fn unitless_and_mm_do_not_warn() {
        assert_eq!(factor_to_mm(0), (1.0, false));
        assert_eq!(factor_to_mm(4), (1.0, false));
    }

    #[test]
    fn codes_that_were_unknown_now_convert() {
        assert_eq!(factor_to_mm(3), (1_609_344.0, false));
        assert_eq!(factor_to_mm(7), (1.0e6, false));
        assert_eq!(factor_to_mm(20), (3.085_677_581_491_367_3e19, false));
    }

    #[test]
    fn beyond_the_table_still_warns_and_assumes_mm() {
        // 21-24 = unités « US survey » : hors table $INSUNITS 0-20 du lot 2b,
        // ezdxf n'a pas de facteur pour elles non plus.
        assert_eq!(factor_to_mm(21), (1.0, true));
        assert_eq!(factor_to_mm(99), (1.0, true));
        assert_eq!(factor_to_mm(-1), (1.0, true));
        assert_eq!(unit_name(99), "unknown");
    }

    #[test]
    fn implausible_units_are_flagged_but_still_exact() {
        // km, Mm, années-lumière : converties exactement, et au-dessus du
        // seuil qui déclenche l'avertissement.
        for code in [3, 7, 15, 16, 17, 18, 19, 20] {
            let (f, unknown) = factor_to_mm(code);
            assert!(!unknown);
            assert!(f >= IMPLAUSIBLE_FACTOR_MM, "code {code} : facteur {f}");
        }
        // Les unités d'atelier restent sous le seuil.
        for code in [1, 2, 4, 5, 6, 9, 10, 13, 14] {
            assert!(factor_to_mm(code).0 < IMPLAUSIBLE_FACTOR_MM, "code {code}");
        }
    }
}
